### PR TITLE
Update to ExoPlayer 2.12.3

### PR DESCRIPTION
**Pull request by Redirion**
#### What is it?
- [ ] Bugfix (user facing)
- [ ] Feature (user facing)
- [X] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
Updates ExoPlayer from 2.11 to 2.12

Note that we will have to do another PR to migrate from deprecated mediasession to media2.
However this will most likely require to also rework everything to the new MediaItems instead of working with MediaSources.
For now however this can stay and if ExoPlayer stays being compatible with mediasession we aren't in a rush.

#### Due diligence
- [X] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by TacoTheDank** [comment]
@Redirion Also, I think media2 requires API 21 or newer, so that would require dropping KitKat support, which could perhaps cut off a small chunk of users :thinking:

**Comment by Stypox** [comment]
@TacoTheDank we will be forced to drop KitKat support sooner or later anyway due to OkHttp being discontinued on [December 31, 2021](https://github.com/square/okhttp#requirements)

**Comment by TacoTheDank** [comment]
@Stypox All okay with me lol, though I do feel kinda bad for those older phone users. The only option they really have left now is SkyTube or using a mobile browser. (Also, you should edit your comment to specify the 3.12.x LTS branch, because you scared me with "OkHttp being discontinued" 😰)

**Comment by Redirion** [comment]
@TacoTheDank do you have a source for the API 21 requirement of media2 for me? I wasn't able to find that info.

**Comment by TacoTheDank** [comment]
@Redirion Nevermind, I think I misunderstood something. I got that idea from the answer here: https://stackoverflow.com/questions/57566102/what-are-the-differences-between-mediasessioncompat-and-mediasession-in-android and thought it meant that Media2 would only support API 21 and newer

oops :sweat_smile:

**Comment by Redirion** [comment]
found something. According to this website media2 in the current version that is used by ExoPlayer requires API 16: https://androidx.tech/artifacts/media2/media2-session/1.1.0

while the media2 extension of ExoPlayer declares API 19: https://github.com/google/ExoPlayer/blob/release-v2/extensions/media2/build.gradle (this could however be overwritten by our gradle config, if ExoPlayer raises this to 21, as long as media2 stays below 21).

**Comment by TacoTheDank** [comment]
@Redirion I see... :thinking: 

Oh btw, Exoplayer 2.12.0 introduced two convenience methods `play()` and `pause()`, which are equivalent to `setPlayWhenReady(true)` and `setPlayWhenReady(false)`, respectively. There are two usages of the latter set in https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/player/helper/AudioReactor.java that could be replaced with the former. ([Here](https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/player/helper/AudioReactor.java)'s the link)

**Comment by Stypox** [comment]
> The only option they really have left now is SkyTube or using a mobile browser

@TacoTheDank they would still be able to use NewPipe legacy

**Comment by opusforlife2** [comment]
@Stypox Does Newpipe Legacy not use OkHttp?

**Comment by Stypox** [comment]
It does, but I think it can be kept or replaced with something else specific to old android versions

**Comment by TacoTheDank** [comment]
> The only option they really have left now is SkyTube or using a mobile browser

Oh yeah, and that :)

**Comment by TacoTheDank** [comment]
Is this good to merge?

**Comment by Redirion** [comment]
@Stypox I am a bit hesitant to merge my own PR. From my opinion I would say this is good to merge. Or would you prefer if we skip ExoPlayer 2.12 and move directly to 2.13?

**Comment by Stypox** [comment]
I added a commit with a few style improvements and fixes of warnings. I tested playing some videos and enabling skip silence and everything worked as expected

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 1]
```diff
@@ -3358,7 +3361,7 @@ private void onTextTracksChanged() {
         final List<String> availableLanguages = new ArrayList<>(textTracks.length);
         for (int i = 0; i < textTracks.length; i++) {
             final TrackGroup textTrack = textTracks.get(i);
-            if (textTrack.length > 0 && textTrack.getFormat(0) != null) {
```
Comment (Redirion) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:3358]: what was the warning here? can never be null?

Comment (Stypox) [reply app/src/main/java/org/schabi/newpipe/player/Player.java:3358]: Yes, probably a `@NonNull` was added by exo player

**Comment by Redirion** [comment]
Now we seem to have the situation, that we require a third dev to approve this PR as we both contribute to it

**Comment by opusforlife2** [comment]
You could always approve each other's changes. :P

**Comment by opusforlife2** [comment]
@Redirion Were you ever invited to join the Newpipe channel on IRC? If not, could you please make a Matrix account and join it?

**Comment by Redirion** [comment]
> You could always approve each other's changes. :P

I just tried but I may not approve my own pull request even though the last commit was not by me

**Review by Stypox — APPROVED** [review]
I tested again and everything works fine. Thanks :-)

**Comment by opusforlife2** [comment]
> Were you ever invited to join the Newpipe channel on IRC? If not, could you please make a Matrix account and join it?

@Redirion

**Pull request merged**